client/jmx-adapter/src/main/java/org/jolokia/client/jmxadapter/RemoteJmxAdapter.java

```diff
@@ -0,0 +1,586 @@
+
+    for (String attribute : attributes) {
+      requests.add(new J4pReadRequest(name, attribute));
+    }
+    List<?> responses = Collections.emptyList();
+    try {
+      responses = this.connector.execute(requests, this.defaultProcessingOptions());
+
+    } catch (J4pBulkRemoteException e) {
+      responses = e.getResults();
+    } catch (J4pException ignore) {
+      //will result in empty return
+    }
+    for (Object item : responses) {
+      if (item instanceof J4pReadResponse) {
+        J4pReadResponse value = (J4pReadResponse) item;
+        final String attribute = value.getRequest().getAttribute();
+        result.add(new Attribute(attribute,
+            adaptJsonToOptimalResponseValue(name, attribute, value.getValue())));
+      }
+    }
+    return result;
+  }
+
+  @Override
+  public void setAttribute(ObjectName name, Attribute attribute)
+      throws InstanceNotFoundException, AttributeNotFoundException, InvalidAttributeValueException,
+      IOException {
+    final J4pWriteRequest request =
+        new J4pWriteRequest(name, attribute.getName(), attribute.getValue());
+    try {
+      this.unwrapExecute(request);
+    } catch (UncheckedJmxAdapterException e) {
+      if (e.getCause() instanceof J4pRemoteException) {
+        J4pRemoteException remote = (J4pRemoteException) e.getCause();
+        if ("javax.management.AttributeNotFoundException".equals(remote.getErrorType())) {
+          throw new AttributeNotFoundException((e.getCause().getMessage()));
+        }
+        if ("java.lang.IllegalArgumentException".equals(remote.getErrorType())
+            && remote
+            .getMessage()
+            .matches(
+                "Error: java.lang.IllegalArgumentException : Invalid value .+ for attribute .+")) {
+          throw new InvalidAttributeValueException(remote.getMessage());
+        }
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public AttributeList setAttributes(ObjectName name, AttributeList attributes)
+      throws InstanceNotFoundException, IOException {
+
+    List<J4pWriteRequest> attributeWrites = new ArrayList<J4pWriteRequest>(attributes.size());
+    for (Attribute attribute : attributes.asList()) {
+      attributeWrites.add(new J4pWriteRequest(name, attribute.getName(), attribute.getValue()));
+    }
+    try {
+      this.connector.execute(attributeWrites);
+    } catch (J4pException e) {
+      unwrapException(e);
+    }
+
+    return attributes;
+  }
+
+  @Override
+  public Object invoke(ObjectName name, String operationName, Object[] params, String[] signature)
+      throws InstanceNotFoundException, MBeanException, IOException {
+    //jvisualvm may send null for no parameters
+    if (params == null && signature.length == 0) {
+      params = new Object[0];
+    }
+    try {
+      final J4pExecResponse response =
+          unwrapExecute(
+              new J4pExecRequest(name, operationName + makeSignature(signature), params));
```

Comment:
Note : For JMC flight recordings we need to support TabularData as parameters. Not sure if that should be put in here, or if there is unlikely that it will be required elsewhere.  https://github.com/skarsaune/jmc-cloud-extensions/blob/b01e4be32436f9297f49def2523577a30028223d/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JmcJolokiaJmxConnection.java#L89
Alternatively it could be handled in the java client library: https://github.com/rhuss/jolokia/blob/19ae180dae49c46ea13dff57a04af3819c8eb324/client/java/src/main/java/org/jolokia/client/request/J4pRequest.java#L207